File: CSharp/XamarinXamlPreviewCS/XamarinXamlPreviewCS/XamarinXamlPreviewCS/Libs/Binding.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Input;

namespace XamarinXamlPreviewCS.Libs
{
    public class ViewModelBase : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler PropertyChanged;
        protected bool SetProperty<T>(ref T storage, T value, [CallerMemberName] String propertyName = null)
        {
            if (object.Equals(storage, value)) return false;

            storage = value;
            this.OnPropertyChanged(propertyName);
            return true;
        }
        protected void OnPropertyChanged([CallerMemberName] string propertyName = null)
        {
            var eventHandler = this.PropertyChanged;
            if (eventHandler != null)
            {
                eventHandler(this, new PropertyChangedEventArgs(propertyName));
            }
        }
    }

    public class DelegateCommand : ICommand
    {
        protected Func<object, bool> canExecute;
        protected Action<object> execute;
        public event EventHandler CanExecuteChanged = (s, e) => { };

        public DelegateCommand(Func<object, bool> can, Action<object> exec)
        {
            canExecute = can;
            execute = exec;
        }
        public bool CanExecute(object parameter)
        {
            if (canExecute != null)
            {
                return canExecute(parameter);
            }
            else
            {
                return false;
            }
        }

        public void Execute(object parameter)
        {
            if (execute != null)
            {
                execute(parameter);
            }
        }
    }
}

File: XamarinXamlPrev/Libs/Binding.fs
namespace XamarinXamlPrev
open System
open System.Collections.Generic
open System.Linq
open System.Text
open Xamarin.Forms
open Moonmile.XFormsProvider
open System.IO
open System.Reflection
open System.ComponentModel
open System.Windows.Input

[<AllowNullLiteral>]
type ViewModelBase() =
    let propertyChangedEvent = new DelegateEvent<PropertyChangedEventHandler>()
    interface INotifyPropertyChanged with
        [<CLIEvent>]
        member x.PropertyChanged = propertyChangedEvent.Publish
    member x.OnPropertyChanged propertyName = 
        propertyChangedEvent.Trigger([| x; new PropertyChangedEventArgs(propertyName) |])

type DelegateCommand(can, exec) =
    let canExecute:(obj -> bool) = can
    let execute:(obj -> unit ) = exec
    let eh = new DelegateEvent<EventHandler> ()
    interface ICommand with
        [<CLIEvent>]
        member x.CanExecuteChanged = eh.Publish
        member x.CanExecute(parameter: obj): bool = canExecute( parameter )
        member x.Execute(parameter: obj): unit = execute(parameter)
